src/main/java/org/jabref/logic/ai/embeddings/EmbeddingsGenerationTaskManager.java

```diff
@@ -0,0 +1,202 @@
+        titleProperty().set(Localization.lang("Embeddings generation"));
+
+        this.onFailure(e -> {
+            throw new RuntimeException(e);
+        });
+    }
+
+    private void listenToPreferences() {
+        // TODO: Will these listeners align with AiIngestor preference listeners?
+        aiService.getPreferences().onEmbeddingsParametersChange(this::invalidate);
+    }
+
+    public void addToProcess(Collection<BibEntry> bibEntries) {
+        bibEntries.forEach(this::addToProcess);
+    }
+
+    public void removeFromProcess(Collection<BibEntry> bibEntries) {
+        bibEntries.forEach(this::removeFromProcess);
+    }
+
+    public void addToProcess(BibEntry bibEntry) {
+        addToProcess(bibEntry.getFiles());
+    }
+
+    public void removeFromProcess(BibEntry bibEntry) {
+        removeFromProcess(bibEntry.getFiles());
+    }
+
+    public void addToProcess(List<LinkedFile> linkedFiles) {
+        linkedFiles.forEach(this::addToProcess);
+    }
+
+    public void removeFromProcess(List<LinkedFile> linkedFiles) {
+        linkedFiles.forEach(this::removeFromProcess);
+    }
+
+    public void addToProcess(LinkedFile linkedFile) {
+        if (!isBlockingNewTasks) {
+            linkedFileQueue.add(linkedFile);
+
+            synchronized (lock) {
+                if (!isRunning) {
+                    this.executeWith(taskExecutor);
+                    showToUser(false);
+                }
+            }
+        }
+    }
+
+    public AutoCloseable blockNewTasks() {
+        synchronized (lock) {
+            isBlockingNewTasks = true;
+        }
+
+        return () -> {
+            synchronized (lock) {
+                isBlockingNewTasks = false;
+            }
+        };
+    }
+
+    public void removeFromProcess(LinkedFile linkedFile) {
+        aiService.getEmbeddingsManager().removeIngestedFile(linkedFile.getLink());
+    }
+
+    public void removeFromProcess(Set<String> linksToRemove) {
+        linksToRemove.forEach(this::removeFromProcess);
+    }
+
+    public void removeFromProcess(String link) {
+        aiService.getEmbeddingsManager().removeIngestedFile(link);
+    }
+
+    public void updateEmbeddings(BibDatabaseContext bibDatabaseContext) {
+        Set<String> linksToRemove = aiService.getEmbeddingsManager().getIngestedFilesTracker().getListOfIngestedFilesLinks();
+        bibDatabaseContext.getEntries().stream()
+                       .flatMap(entry -> entry.getFiles().stream())
+                       .map(LinkedFile::getLink)
+                       .forEach(linksToRemove::remove);
+
+        removeFromProcess(linksToRemove);
+
+        addToProcess(bibDatabaseContext.getEntries());
+    }
+
+    @Override
+    protected Void call() throws Exception {
+        synchronized (lock) {
+            isRunning = true;
+        }
+
+        updateProgress();
+
+        while (!linkedFileQueue.isEmpty() && !isCanceled()) {
+            LinkedFile linkedFile = linkedFileQueue.removeLast();
```

Comment:
Typically, queues are LIFO, thus one does `.removeFirst()`

Reply:
But isn't `.removeLast()` is O(1) (except the case where array is grown and reallocated) and `.removeFirst()` is O(n) for an `ArrayList`?